// @mantine/tiptap/esm/RichTextEditorControl/RichTextEditorColorPickerControl.mjs
'use client';
import React, { forwardRef, useState } from 'react';
import { useProps, ColorSwatch, Popover, SimpleGrid, ColorPicker, Tooltip, Group, ActionIcon, rem } from '@mantine/core';
import { useDisclosure } from '@mantine/hooks';
import { IconX, IconCircleOff, IconColorPicker, IconPalette, IconCheck } from '../icons/Icons.mjs';
import { useRichTextEditorContext } from '../RichTextEditor.context.mjs';
import { RichTextEditorControl } from './RichTextEditorControl.mjs';

const defaultProps = {};
const RichTextEditorColorPickerControl = forwardRef((props, ref) => {
  const { popoverProps, colors, colorPickerProps, ...others } = useProps(
    "RichTextEditorColorPickerControl",
    defaultProps,
    props
  );
  const { editor, labels, getStyles } = useRichTextEditorContext();
  const [opened, { toggle, close }] = useDisclosure(false);
  const [state, setState] = useState("palette");
  const currentColor = editor?.getAttributes("textStyle").color || "var(--mantine-color-text)";
  const handleChange = (value, shouldClose = true) => {
    (editor?.chain()).focus().setColor(value).run();
    shouldClose && close();
  };
  const handleClear = () => {
    (editor?.chain()).focus().unsetColor().run();
    close();
  };
  const controls = colors.map((color, index) => /* @__PURE__ */ React.createElement(
    ColorSwatch,
    {
      key: index,
      component: "button",
      color,
      onClick: () => handleChange(color),
      size: 26,
      radius: "xs",
      style: { cursor: "pointer" },
      title: labels.colorPickerColorLabel(color),
      "aria-label": labels.colorPickerColorLabel(color)
    }
  ));
  return /* @__PURE__ */ React.createElement(Popover, { opened, withinPortal: true, trapFocus: true, onClose: close, ...popoverProps }, /* @__PURE__ */ React.createElement(Popover.Target, null, /* @__PURE__ */ React.createElement(
    RichTextEditorControl,
    {
      ...others,
      "aria-label": labels.colorPickerControlLabel,
      title: labels.colorPickerControlLabel,
      ref,
      onClick: toggle
    },
    /* @__PURE__ */ React.createElement(ColorSwatch, { color: currentColor, size: 14 })
  )), /* @__PURE__ */ React.createElement(Popover.Dropdown, { ...getStyles("linkEditorDropdown") }, state === "palette" && /* @__PURE__ */ React.createElement(SimpleGrid, { cols: 7, spacing: 2 }, controls), state === "colorPicker" && /* @__PURE__ */ React.createElement(
    ColorPicker,
    {
      defaultValue: currentColor,
      onChange: (value) => handleChange(value, false),
      ...colorPickerProps
    }
  ), /* @__PURE__ */ React.createElement(Tooltip.Group, { closeDelay: 200 }, /* @__PURE__ */ React.createElement(Group, { justify: "flex-end", gap: "xs", mt: "sm" }, state === "palette" && /* @__PURE__ */ React.createElement(
    ActionIcon,
    {
      variant: "default",
      onClick: close,
      title: labels.colorPickerCancel,
      "aria-label": labels.colorPickerCancel
    },
    /* @__PURE__ */ React.createElement(IconX, { style: { width: rem(16), height: rem(16) } })
  ), /* @__PURE__ */ React.createElement(
    ActionIcon,
    {
      variant: "default",
      onClick: handleClear,
      title: labels.colorPickerClear,
      "aria-label": labels.colorPickerClear
    },
    /* @__PURE__ */ React.createElement(IconCircleOff, { style: { width: rem(16), height: rem(16) } })
  ), state === "palette" ? /* @__PURE__ */ React.createElement(
    ActionIcon,
    {
      variant: "default",
      onClick: () => setState("colorPicker"),
      title: labels.colorPickerColorPicker,
      "aria-label": labels.colorPickerColorPicker
    },
    /* @__PURE__ */ React.createElement(IconColorPicker, { style: { width: rem(16), height: rem(16) } })
  ) : /* @__PURE__ */ React.createElement(
    ActionIcon,
    {
      variant: "default",
      onClick: () => setState("palette"),
      "aria-label": labels.colorPickerPalette,
      title: labels.colorPickerPalette
    },
    /* @__PURE__ */ React.createElement(IconPalette, { style: { width: rem(16), height: rem(16) } })
  ), state === "colorPicker" && /* @__PURE__ */ React.createElement(
    ActionIcon,
    {
      variant: "default",
      onClick: close,
      title: labels.colorPickerSave,
      "aria-label": labels.colorPickerSave
    },
    /* @__PURE__ */ React.createElement(IconCheck, { style: { width: rem(16), height: rem(16) } })
  )))));
});
RichTextEditorColorPickerControl.displayName = "@mantine/tiptap/ColorPickerControl";

export { RichTextEditorColorPickerControl };
